packages/url_launcher/url_launcher_web/lib/link.dart

```diff
@@ -0,0 +1,295 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:html' as html;
+import 'dart:js_util';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/gestures.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter/services.dart';
+import 'package:flutter_web_plugins/flutter_web_plugins.dart';
+
+import 'package:url_launcher_platform_interface/link.dart';
+
+typedef _ClickListener = void Function(html.MouseEvent);
+
+/// The unique identifier for the view type to be used for link platform views.
+const String viewType = '__url_launcher::link';
+
+/// The name of the property used to set the viewId on the DOM element.
+const String _viewIdProperty = '__url_launcher::link::viewId';
+
+// TODO(mdebbar): Should use PlatformViewFactory.
+dynamic get webViewFactory => _LinkViewController._viewFactory;
+
+final MethodCodec _codec = const JSONMethodCodec();
+
+/// The delegate for building the [Link] widget on the web.
+///
+/// It uses a platform view to render an anchor element in the DOM.
+class WebLinkDelegate extends StatefulWidget {
+  /// Creates a delegate for the given [link].
+  const WebLinkDelegate(this.link);
+
+  /// The [Link] widget built by the app.
+  final Link link;
+
+  @override
+  _WebLinkDelegateState createState() => _WebLinkDelegateState();
+}
+
+class _WebLinkDelegateState extends State<WebLinkDelegate> {
+  _LinkViewController _controller;
+
+  @override
+  void didUpdateWidget(WebLinkDelegate oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.link.uri != oldWidget.link.uri) {
+      _controller?._setUri(widget.link.uri);
+    }
+    if (widget.link.target != oldWidget.link.target) {
+      _controller?._setTarget(widget.link.target);
+    }
+  }
+
+  Future<void> _followLink() {
+    final Completer<void> completer = Completer<void>();
+    _LinkViewController._registerHitTest(
+      _controller,
+      onClick: (html.MouseEvent event) {
+        completer.complete(_onDomClick(event));
+      },
+    );
+    return completer.future;
+  }
+
+  Future<void> _onDomClick(html.MouseEvent event) {
+    if (!widget.link.uri.hasScheme) {
+      // A uri that doesn't have a scheme is an internal route name. In this
+      // case, we push it via Flutter's navigation system instead of letting the
+      // browser handle it.
+      event.preventDefault();
+      final String routeName = widget.link.uri.toString();
+      // TODO(mdebbar): how do we know if this is true or false?
+      final bool isUsingRouter = false;
+      // QUESTION: does this update the history entry on the engine side?
+      final MethodCall call = isUsingRouter
+          ? MethodCall('pushRouteInformation', <dynamic, dynamic>{
+              'location': routeName,
+              'state': null,
+            })
+          : MethodCall('pushRoute', routeName);
```

Comment:
@chunhtai is there a way currently to tell if the app is in Router mode? Maybe inspect the widget hierarchy for a Router widget?

Reply:
there is no standard API to check that. Since this is a separate package, i think just do a Router.of(context, nullOk = true) != null should be fine

Reply:
Thanks!